lib/active_resource/threadsafe_attributes.rb

```diff
@@ -0,0 +1,57 @@
+module ThreadsafeAttributes
+  def self.included(klass)
+    klass.extend(ClassMethods)
+  end
+
+  module ClassMethods
+    def threadsafe_attribute(*attrs)
+      attrs.each do |attr|
+        define_method attr do
+          get_threadsafe_attribute(attr)
+        end
+
+        define_method "#{attr}=" do |value|
+          set_threadsafe_attribute(attr, value)
+        end
+
+        define_method "#{attr}_defined?" do
+          threadsafe_attribute_defined?(attr)
+        end
+      end
+    end
+  end
+
+  private
+
+  def get_threadsafe_attribute(name)
+    if threadsafe_attribute_defined_by_thread?(name, Thread.current)
+      get_threadsafe_attribute_by_thread(name, Thread.current)
+    elsif threadsafe_attribute_defined_by_thread?(name, Thread.main)
+      get_threadsafe_attribute_by_thread(name, Thread.main).dup.tap do |value|
```

Comment:
Depending on the object the `dup` here may not work since it is not going to dup the deep objects like hashes and arrays. Do we need to take care of this case?